worldengine/simulations/icecap.py

```diff
@@ -0,0 +1,79 @@
+import numpy
+
+
+class IcecapSimulation(object):
+    # This class creates an "ice-map", i.e. a numpy array with positive values that describe the thickness of the ice at
+    # a certain spot of the world.
+    # Ice can appear wherever there is an ocean and the temperature is cold enough.
+
+    # TODO: Find out if a desert planet could still freeze or if the freeze-threshold is dynamic.
+    # TODO: Freeze rivers etc.
+
+    @staticmethod
+    def is_applicable(world):
+        return world.has_ocean() and world.has_temperature()
+
+    def execute(self, world, seed):
+        world.icecap = self._calculate(world, seed)
+
+    @staticmethod
+    def _calculate(world, seed):
+        # Notes on performance:
+        #  -method is run once per generation
+        #  -iterations        : width * height
+        #  -memory consumption: width * height * sizeof(numpy.float) (permanent)
+        #                       width * height * sizeof(numpy.bool) (temporary)
+
+        # constants for convenience (or performance)
+        ocean = world.ocean
+        temperature = world.temperature['data']
+
+        # primary constants (could be used as global variables at some point); all values should be in [0, 1]
+        max_freeze_percentage = 0.60  # only the coldest x% of the cold area will freeze (0 = no ice, 1 = all ice)
+        freeze_chance_window = 0.20  # the warmest x% of freezable area won't completely freeze (RNG decides)
+        surrounding_tile_influence = 0.5  # chance-modifier to freeze a slightly warm tile when neighbors are frozen
+
+        # secondary constants
+        temp_min = temperature.min()  # coldest spot in the world
+        freeze_threshold = world.temperature['thresholds'][0][1]  # upper temperature-limit for freezing effects
+        # Cold biomes: TODO: find and pick most appropriate threshold
+        #    polar: self.temperature['thresholds'][0][1]
+        #   alpine: self.temperature['thresholds'][1][1]
+        #   boreal: self.temperature['thresholds'][2][1]
+
+        # derived constants
+        freeze_threshold = (freeze_threshold - temp_min) * max_freeze_percentage  # calculate freeze threshold above min
+        freeze_chance_threshold = freeze_threshold * (1.0 - freeze_chance_window)
+
+        # local variables
+        icecap = numpy.zeros((world.height, world.width), dtype=float)
+        rng = numpy.random.RandomState(seed)  # create our own random generator
+
+        # map that is True whenever there is land or (certain) ice around
+        solid_map = numpy.logical_or(temperature <= freeze_chance_threshold + temp_min, numpy.logical_not(ocean))
+
+        for y in range(world.height):
+            for x in range(world.width):
+                if ocean[y, x]:  # or world.river_map[y, x] > 0 or world.lake_map[y, x] > 0 or world.watermap['data'][y, x] > 0:
```

Comment:
There could be more checks, the code itself is not dependent on there being ocean. But I wouldn't know how to properly handle other freezable places.